### PR TITLE
Remove gotos from DFU Init

### PR DESCRIPTION
**Pull request by PML-Trey**
Gotos were being used to control DFU Device init.  Gotos are bad programming practice and made the device initialization software convoluted.  This commit removes these gotos and breaks device initialization into functions with a more straightforward execution flow.

### CODE REVIEW

File: src/dfu.c [hunk 1]
```diff
@@ -441,10 +485,10 @@ struct usb_device *dfu_device_init( const uint32_t vendor,
         for( usb_bus = usb_get_busses(); NULL != usb_bus; usb_bus = usb_bus->next ) {
             for( device = usb_bus->devices; NULL != device; device = device->next) {
                 if(    (vendor  == device->descriptor.idVendor)
-                    && (product == device->descriptor.idProduct)
-                    && ((bus_number == 0)
-                        || (device->devnum == device_address
-                            && (usb_bus->location >> 24) == bus_number)))
+                       && (product == device->descriptor.idProduct)
```
Comment (rebpdx) [inline src/dfu.c]: Is this making their code conform to their standard or changing their standard?

File: src/dfu.c [hunk 2]
```diff
@@ -510,39 +554,39 @@ char* dfu_state_to_string( const int32_t state ) {
...
-        case STATE_DFU_MANIFEST_WAIT_RESET:
-            message = "dfuMANIFEST-WAIT-RESET";
-            break;
-        case STATE_DFU_UPLOAD_IDLE:
-            message = "dfuUPLOAD-IDLE";
-            break;
-        case STATE_DFU_ERROR:
-            message = "dfuERROR";
-            break;
+    case STATE_APP_IDLE:
```
Comment (rebpdx) [inline src/dfu.c]: Same question fixing code standard or changing it?

File: src/dfu.c [hunk 3]
```diff
@@ -635,30 +689,37 @@ static int32_t dfu_find_interface( struct libusb_device *device,
...
                 }
             }
         }
 
         libusb_free_config_descriptor( config );
     }
 
-    return -1;
+    libusb_close(dfu_device->handle);
+    DEBUG( "Closing device\n");
```
Comment (rebpdx) [inline src/dfu.c]: Spacing for both of these functions?

File: src/dfu.c [hunk 4]
```diff
@@ -552,63 +596,73 @@ char* dfu_status_to_string( const int32_t status ) {
...
 static int32_t dfu_find_interface( struct libusb_device *device,
+                                   dfu_device_t *dfu_device,
                                    const dfu_bool honor_interfaceclass,
                                    const uint8_t bNumConfigurations) {
     int32_t c,i,s;
+    int32_t dfuInterface = -1;
+
+    int debug;
+
+    debug = libusb_open(device, &dfu_device->handle);
```
Comment (rebpdx) [inline src/dfu.c]: Spacing?

File: src/dfu.c [hunk 5]
```diff
@@ -314,100 +314,138 @@ int32_t dfu_abort( dfu_device_t *device ) {
...
+    while(retries)
     {
-        usleep(500000);
-        goto retry;
+        //Find a device with a matching VID/PID
+        libusb_device *device = NULL;
+        while((device == NULL) && (retries > 0))
+        {
+            DEBUG( "Finding device retry #: %d\n", retries );
+            device = dfu_find_device(list, &descriptor, vendor, product, bus_number, device_address);
```
Comment (rebpdx) [inline src/dfu.c]: Function call spacing?

File: src/dfu.c [hunk 6]
```diff
@@ -314,100 +314,138 @@ int32_t dfu_abort( dfu_device_t *device ) {
...
+        }
+        //Leave if we didn't find a device
+        if(device == NULL)
+            return NULL;
+
+        //Try to open the device
+        DEBUG(" Calling dfu_open_device\n");
+        if(dfu_open_device(device, dfu_device, &descriptor, honor_interfaceclass))
+        {
+            if(0 == dfu_make_idle(dfu_device, initial_abort))
```
Comment (rebpdx) [inline src/dfu.c]: Function call spacing?

File: src/dfu.c [hunk 7]
```diff
@@ -314,100 +314,138 @@ int32_t dfu_abort( dfu_device_t *device ) {
...
+            //Wait 100 ms before looking again
+            usleep(100000);
+        }
+        //Leave if we didn't find a device
+        if(device == NULL)
+            return NULL;
+
+        //Try to open the device
+        DEBUG(" Calling dfu_open_device\n");
+        if(dfu_open_device(device, dfu_device, &descriptor, honor_interfaceclass))
```
Comment (rebpdx) [inline src/dfu.c]: Function call spacing?

File: src/dfu.c [hunk 8]
```diff
@@ -314,100 +314,138 @@ int32_t dfu_abort( dfu_device_t *device ) {
...
+            retries--;
+            //Wait 100 ms before looking again
+            usleep(100000);
+        }
+        //Leave if we didn't find a device
+        if(device == NULL)
+            return NULL;
+
+        //Try to open the device
+        DEBUG(" Calling dfu_open_device\n");
```
[the remark this answers is not in the document]
Comment (PML-Trey) [reply src/dfu.c]: Space can be removed

File: src/dfu.c [hunk 9]
```diff
@@ -314,100 +314,138 @@ int32_t dfu_abort( dfu_device_t *device ) {
...
+            else
+            {
+                //Device was either reset or couldn't get into DFUIdle so try again
+                libusb_free_device_list( list, 1 );
+                usleep(100000);
+                retries--;
+                continue;
+            }
+
+        }
```
[the remark this answers is not in the document]
Comment (PML-Trey) [reply src/dfu.c]: They don't really have a standard...

File: src/dfu.c [hunk 10]
```diff
@@ -416,6 +454,12 @@ struct libusb_device *dfu_device_init( const uint32_t vendor,
 
     return NULL;
 }
+
+#endif
```
[the remark this answers is not in the document]
Comment (PML-Trey) [reply src/dfu.c]: Yes, good idea

File: src/dfu.c (hunk 1 above)
Comment (PML-Trey) [reply src/dfu.c]: There isn't much of a standard in their code.  Looks like the code formatter changed the white space a bit

File: src/dfu.c (hunk 2 above)
Comment (PML-Trey) [reply src/dfu.c]: They indent cases.  Code formatter changed it.

File: src/dfu.c [hunk 11]
```diff
@@ -314,100 +314,138 @@ int32_t dfu_abort( dfu_device_t *device ) {
...
+    extern libusb_context *usbcontext;
 
-    devicecount = libusb_get_device_list( usbcontext, &list );
+    device_count = libusb_get_device_list( usbcontext, &list );
 
-    for( i = 0; i < devicecount; i++ ) {
+    for( i = 0; i < device_count; i++ ) {
         libusb_device *device = list[i];
-        struct libusb_device_descriptor descriptor;
+        //struct libusb_device_descriptor descriptor;
```
[the remark this answers is not in the document]
Comment (PML-Trey) [reply src/dfu.c]: Oops

File: src/dfu.c (hunk 5 above)
Comment (PML-Trey) [reply src/dfu.c]: Can be changed

File: src/dfu.c (hunk 7 above)
Comment (PML-Trey) [reply src/dfu.c]: copy that

**Pull request merged**